synapse/handlers/room_batch.py

```diff
@@ -176,7 +177,7 @@ async def persist_state_events_at_start(
         assert app_service_requester.app_service
 
         state_event_ids_at_start = []
-        auth_event_ids = initial_auth_event_ids.copy()
+        state_event_ids = initial_state_ids_at_event.copy()
```

Comment:
(Side note: all these copies will be expensive for rooms with a lot of state, which will affect how quickly we can batch insert history.)

Reply:
Maybe we can get rid of this `copy()` since we `copy()` below again. But we do need to `copy()` in the other spots since I think there are areas where it's modified in place.

Reply:
Nevermind, we do modify it in the loop here and also copy to pass a snapshot of it

Reply:
I'm not sure how exactly to side-step this.

`state_event_ids` is an ongoing list where we add to as we persist `state_events_at_start`. Then return the full list at the end.

And we `state_event_ids.copy()` to snapshot the state in the loop and pass it to the event creation functions. Otherwise, the `signatures` integrity checks go out of whack because we modified the list in place.

Reply:
For reference on why we have to define `state_event_ids` here:

```
# Since each state event is marked as an outlier, the
# `EventContext.for_outlier()` won't have any `state_ids`
# set and therefore can't derive any state even though the
# prev_events are set. Also since the first event in the
# state chain is floating with no `prev_events`, it can't
# derive state from anywhere automatically. So we need to
# set some state explicitly.
```